lec08-stability.Rmd

```diff
@@ -1,4 +1,589 @@
+    ggplot(aes(x = time, y = Value, color = Column)) +
+    geom_line(aes(x = time, y = Value)) +
+    labs(y = "Population size")
+```
+
+This plot very nicely summarizes the behaviour of the logistic equation: it shows the
+path that the population size $N$ will take in time from several initial conditions. 
+Importantly, the carrying capacity $K$ which we set to be $50$ is clearly a stable 
+fixed point: all the trajectories go towards $K$. 
+
+#### Challenge
+
+Modify the plotting code above so that the legend lists the initial conditions for each
+line instead of `X1`, `X2`, etc. 
+
+```{r, eval=FALSE, echo=FALSE}
+# I don't know how to do this
+
+```
+
+## Qualitative analysis of 2D models
+
+Now we'll move on to look at how to analyze two-dimensional models. **Two-dimensional** 
+means that there are now two variables in the system (like $x$ and $y$, prey and predators) 
+instead of one (like $N$).
+
+Many of the concepts and tools from analyzing one-dimensional models will be used 
+here as well, but we will add some things that are specific to more than one dimension. 
+
+To motivate some of these new concepts and the idea of modelling in two dimensions, 
+let's analyze a model of two species interacting.
+
+$$\frac{dx}{dt} = ax - bxy$$
+$$\frac{dy}{dt} = cx - dy$$
+
+This is very close to the Lotka-Volterra predator-prey model, but the term for predator 
+growth is $cx$ instead of $cxy$. In this model the predators will eat the prey at a 
+constant rate, no matter how many of the predators there are. Make sure you don't use 
+these exact equations when you do the assignment!
+
+First, we'll simulate a trajectory for the two species.
+
+```{r}
+predator_prey <- function(t, y, parameters) {
+  # calculates dx/dt and dy/dt for a predator-prey model
+  
+  # t: time at which to evaluate derivatives
+  # y: vector of system variables (c(X, Y))
+  # parameters: vector of model parameters (c(a, b, c, d))
+  
+  # the arguments need to be named "t", "y", and "parameters", otherwise this won't work with phaseR, a package we will use later
+    
+  #now the state vector y has two elements because we have two species
+  X <- y[1] # prey
+  Y <- y[2] # predators
+  
+  a <- parameters[1]
+  b <- parameters[2]
+  c <- parameters[3]
+  d <- parameters[4]
+    
+  # calculate rate of change
+  dx <- a * X - b * Y * X
+  dy <- c * X - d * Y
+  
+  # return rate of change
+  return(list(c(dx, dy)))
+}
+```
+
+```{r}
+# run the numerical solution
+
+parameters = c(a = 5, b = 1, c = 1, d = 0.2) 
+state <- c(X = 2, Y = 2)
+times <- seq(0, 50, by = 0.01)
+
+result <- ode(y = state, times = times, func = predator_prey, parms = parameters)
+result <- data.frame(result)
+
+```
+
+```{r}
+# plot the results
+
+ggplot(result) +
+  geom_line(aes(x = time, y = X), color = 'red') +
+  geom_line(aes(x = time, y = Y), color = 'blue') +
+  labs(y = "Population size")
+```
+
+It looks like we have oscillations that decay to a fixed point. 
```

Comment:
Can anchor it in a real world intepretation (like the ones from last class), clarifying what it means to have partial (0.5, 1.7, etc) prey/predators?